server/protocols/webadmin/webadmin-data/src/main/java/org/apache/james/webadmin/routes/DomainMappingsRoutes.java

```diff
@@ -0,0 +1,155 @@
+    public DomainMappingsRoutes(final RecipientRewriteTable recipientRewriteTable, final JsonTransformer jsonTransformer) {
+        this.recipientRewriteTable = recipientRewriteTable;
+        this.jsonTransformer = jsonTransformer;
+    }
+
+    @Override
+    public void define(final Service service) {
+        service.get(DOMAIN_MAPPINGS, this::get, jsonTransformer);
+        service.put(SPECIFIC_MAPPING, this::addDomainMapping);
+        service.delete(SPECIFIC_MAPPING, this::removeDomainMapping);
+    }
+
+    @PUT
+    @Path(SPECIFIC_MAPPING_PATH)
+    @ApiOperation(value = "Creating domain mapping between source and destination domains.")
+    @ApiImplicitParams({
+            @ApiImplicitParam(required = true, dataType = "string", name = FROM_DOMAIN, paramType = "path")
+    })
+    @ApiResponses(value = {
+            @ApiResponse(code = HttpStatus.NO_CONTENT_204, message = "Ok"),
+            @ApiResponse(code = HttpStatus.BAD_REQUEST_400, message = "Domain name is invalid"),
+            @ApiResponse(code = HttpStatus.INTERNAL_SERVER_ERROR_500,
+                    message = "Internal server error - Something went bad on the server side.")
+    })
+    public HaltException addDomainMapping(Request request, Response response) {
+        doMapping(request, recipientRewriteTable::addAliasDomainMapping);
+        return halt(HttpStatus.NO_CONTENT_204);
+    }
+
+    @DELETE
+    @Path(SPECIFIC_MAPPING_PATH)
+    @ApiOperation(value = "Removes domain mapping between source and destination domains.")
+    @ApiImplicitParams({
+            @ApiImplicitParam(required = true, dataType = "string", name = FROM_DOMAIN, paramType = "path")
+    })
+    @ApiResponses(value = {
+            @ApiResponse(code = HttpStatus.NO_CONTENT_204, message = "Ok"),
+            @ApiResponse(code = HttpStatus.BAD_REQUEST_400, message = "Domain name is invalid"),
+            @ApiResponse(code = HttpStatus.INTERNAL_SERVER_ERROR_500,
+                    message = "Internal server error - Something went bad on the server side.")
+    })
+    public HaltException removeDomainMapping(Request request, Response response) {
+        doMapping(request, recipientRewriteTable::removeAliasDomainMapping);
+        return halt(HttpStatus.NO_CONTENT_204);
+    }
+
+    @GET
+    @Path(DOMAIN_MAPPINGS)
+    @ApiOperation(value = "Lists all domain mappings.")
+    @ApiResponses(value = {
+            @ApiResponse(code = HttpStatus.OK_200, message = "Domain mappings.", responseContainer = "Map"),
+            @ApiResponse(code = HttpStatus.INTERNAL_SERVER_ERROR_500,
+                    message = "Internal server error - Something went bad on the server side.")
+    })
+    public Map<String, List<String>> get(Request request, Response response) throws RecipientRewriteTableException {
+        return recipientRewriteTable.getAllMappings()
+                .entrySet()
+                .stream()
+                .collect(toImmutableMap(e -> e.getKey().getFixedDomain(),
```

Comment:
`e` is a bit too short as a name. I'd propose to name it `mappingsEntry`

Reply:
What if the mappings contains no `DomainMapping`? Wouldn't it be included in the response with an empty list?

Wouldn't it make more sens to filter them out? (map the value with the select(Mapping.Type.Domain) then filter out empty mappings?)